mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.kt

```diff
@@ -0,0 +1,1114 @@
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.logging.Level
+import java.util.logging.Logger
+import javax.net.ServerSocketFactory
+import javax.net.ssl.SSLContext
+import javax.net.ssl.SSLSocket
+import javax.net.ssl.SSLSocketFactory
+import javax.net.ssl.TrustManager
+import javax.net.ssl.X509TrustManager
+import kotlin.math.min
+
+/**
+ * A scriptable web server. Callers supply canned responses and the server replays them upon request
+ * in sequence.
+ */
+class MockWebServer : ExternalResource(), Closeable {
+  private val requestQueue = LinkedBlockingQueue<RecordedRequest>()
+  private val openClientSockets =
+      Collections.newSetFromMap(ConcurrentHashMap<Socket, Boolean>())
+  private val openConnections =
+      Collections.newSetFromMap(ConcurrentHashMap<Http2Connection, Boolean>())
+  private val requestCount = AtomicInteger()
+  private var bodyLimit = Long.MAX_VALUE
+  private var serverSocketFactory: ServerSocketFactory? = null
+  private var serverSocket: ServerSocket? = null
+  private var sslSocketFactory: SSLSocketFactory? = null
+  private var executor: ExecutorService? = null
+  private var tunnelProxy: Boolean = false
+  private var clientAuth = CLIENT_AUTH_NONE
+  /**
+   * Returns the dispatcher used to respond to HTTP requests.
+   * The default dispatcher is a [QueueDispatcher] but other dispatchers can be configured.
+   *
+   * Sets the dispatcher used to match incoming requests to mock responses. The default dispatcher
+   * simply serves a fixed sequence of responses from a [queue][enqueue]; custom
+   * dispatchers can vary the response based on timing or the content of the request.
+   */
+  var dispatcher: Dispatcher = QueueDispatcher()
+  private var port = -1
+  private var inetSocketAddress: InetSocketAddress? = null
+  private var protocolNegotiationEnabled = true
+  private var protocols = Util.immutableList(Protocol.HTTP_2, Protocol.HTTP_1_1)
+
+  private var started: Boolean = false
+
+  val hostName: String
+    get() {
+      before()
+      return inetSocketAddress!!.address.canonicalHostName
+    }
+
+  @Synchronized override fun before() {
+    if (started) return
+    try {
+      start()
+    } catch (e: IOException) {
+      throw RuntimeException(e)
+    }
+  }
+
+  fun getPort(): Int {
+    before()
+    return port
+  }
+
+  fun toProxyAddress(): Proxy {
+    before()
+    val address = InetSocketAddress(inetSocketAddress!!.address
+        .canonicalHostName, port)
+    return Proxy(Proxy.Type.HTTP, address)
+  }
+
+  fun setServerSocketFactory(serverSocketFactory: ServerSocketFactory) {
+    if (executor != null) {
+      throw IllegalStateException(
+          "setServerSocketFactory() must be called before start()")
+    }
+    this.serverSocketFactory = serverSocketFactory
+  }
+
+  /**
+   * Returns a URL for connecting to this server.
+   *
+   * @param path the request path, such as "/".
+   */
+  fun url(path: String): HttpUrl? {
```

Comment:
please add some `!!` on the resolve call and make this return a non-null `HttpUrl`

Reply:
Fixed.